### DISCUSSION TITLE
HeaderMatchType(s) too restrictive?

### DISCUSSION THREAD

**Discussion by mikhail-aws** [kubernetes-sigs/gateway-api#2576] [General]
**Intro**
Description for context https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HeaderMatchType

> HeaderMatchType specifies the semantics of how HTTP header values should be compared. Valid HeaderMatchType values, along with their conformance levels, are:
> 
>     “Exact” - Core
>     “RegularExpression” - Implementation Specific
> 
> Note that values may be added to this enum, implementations must ensure that unknown values will not cause a crash.
> 
> Unknown values here must result in the implementation setting the Accepted Condition for the Route to status: False, with a Reason of UnsupportedValue.

**Problem**
Should we relax validation on HeaderMatchType? I think Exact and RegularExpression is reasonable baseline. But at the same time implementation might provide more options beyond these 2. 

We are implementing Gateway-API for AWS Lattice, and lattice supports Contains and Prefix options for HeaderTypes ([ref](https://docs.aws.amazon.com/vpc-lattice/latest/APIReference/API_HeaderMatchType.html)). So now we have a dilemma how to approach this. Either create a very limited RegularExpression that will mimic Prefix and Contains, or CRDs. Both do not sound right.

Are there other options? 

My thinking is that HeaderMatchType should not be restricted to these 2 options. The rule should be "At Least Exact and RegularExpression", but not "Only Exact and RegularExpression". Which also fits into second part of the documentation, we should anticipate new and Unknown types:

> Note that values may be added to this enum, implementations must ensure that unknown values will not cause a crash.
> 
> Unknown values here must result in the implementation setting the Accepted Condition for the Route to status: False, with a Reason of UnsupportedValue.

**Comment by tokers** [comment]
> Either create a very limited RegularExpression that will mimic Prefix and Contains, or CRDs. Both do not sound right.

Since the regular expression can mimic the prefix and contains way, what's your concern about using the regular expression? Performance?

**Reply by mikhail-aws** [reply]
Right now we dont support headers regex for various reasons. Contains, Prefix, Exact cover lots of use-cases without need for regex.

**Comment by robscott** [comment]
With every match type we add, we try to ensure there's a path to portability. For example, if a majority of envoy, haproxy, and nginx could support a match type and there was some kind of evidence of widespread usage of those features, that would be a good signal towards inclusion in the API. 

I haven't looked into these specific match types, but if they meet the criteria I mentioned above, I'd say it would be worth a small GEP to add these to the API.

**Reply by mikhail-aws** [reply]
- *Envoy* - supports. Has extensive list of header matchers and includes  "contains_match", "perfix_match", "exact_match"
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-headermatcher

- *Haproxy* - supports. Also has extensive list of matchers. Looks quite  comprehensive. Supports: prefix - "beg", suffix - "end", substring -  "sub".
``` req.hdr(x-header-1) -m beg HeaderValue ```
https://docs.haproxy.org/2.8/configuration.html#7.1.3

- *Nginx* - cant tell. Their documentation does not explicitly cover how to do header based routing. Nginx Gateway API spec says they support  only [Exact match](https://github.com/nginxinc/nginx-gateway-fabric/blob/2fe03e1fac8715702ad2cf30a25c9a7d5f771707/docs/gateway-api-compatibility.md?plain=1#L145)

Does it meet criteria for GEP?

**Reply by robscott** [reply]
Thanks @mikhail-aws! That sounds promising. @pleshakov and/or @brianehlert can you chime in on what NGINX is capable of here?

**Reply by tokers** [reply]
Em, this let me remember the "match expression" in APISIX where people can specify a left value, an operator and right value to compose an expression. The operator can be:

* Equal
* NotEqual
* GreaterThan
* LessThan
* Contains
* Exist
* ......

This is flexible but not all of the implementations support.

**Reply by pleshakov** [reply]
@robscott 

NGINX without any extensions only supports exact or regex match (details: [map](https://nginx.org/en/docs/http/ngx_http_map_module.html#map) directive, [if](https://nginx.org/en/docs/http/ngx_http_rewrite_module.html#if) directive)

The we [implemented](https://www.nginx.com/blog/how-nginx-gateway-fabric-implements-complex-routing-rules/) advanced routing rules matching like headers in our NGINX-based implementation is using [njs](https://nginx.org/en/docs/njs/), which allows running JavaScript code as part of request processing. Considering that, String matching like prefix, suffix, contains can be supported. (details -- https://nginx.org/en/docs/njs/compatibility.html String prototype methods)

**Reply by robscott** [reply]
Thanks @tokers and @pleshakov! It looks like a lot of implementations could support `contains`, `prefix` and `suffix` matches, even if in some cases they were implemented by regex behind the scenes. @mikhail-aws it seems like this could merit a small GEP that both proposed some more common + portable match types as well as a domain-prefixed match type for implementation-specific matches (as @youngnick suggested below). Let us know if we can help with anything.

**Reply by mikhail-aws** [reply]
I will try to find time in next few weeks to get familiar with contribution guideline and GEP scope. If there is another similar size GEP that I can learn from that would be great.

**Comment by youngnick** [comment]
I agree that it would be better to make this more extensible - but if we're going to allow other values, I think we should use the domain-prefixed string pattern, and reserve barewords for Gateway API reserved words.

So, `Exact` and `RegularExpression` are reserved barewords, while `implementation.domain.io/MatchType` would be a domain-prefixed string, and would be allowed at Implementation Specific support.

All other barewords would be reserved for future expansion - with the intent that `implementation.domain.io/MatchType`, if supported across multiple implementations, could eventually become an Extended `MatchType` constant in Gateway API itself, with clear conformance tests to ensure portability.
